File: backend/handlers/fedproxy.go
package handlers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/depo/backend/middleware"
	"github.com/depo/backend/utils"
	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5/pgxpool"
)

// fedHTTPClient never follows redirects, preventing a malicious federation
// server from issuing a 301/302 redirect to an internal address and bypassing
// the SSRF validation that was applied to the original hostname.
var fedHTTPClient = &http.Client{
	Timeout: 30 * time.Second,
	CheckRedirect: func(req *http.Request, via []*http.Request) error {
		return http.ErrUseLastResponse
	},
}

type FedProxyHandler struct {
	DB     *pgxpool.Pool
	AppEnv string
}

type IncomingShare struct {
	ID                     string    `json:"id"`
	RemoteServer           string    `json:"remoteServer"`
	EncryptedCollectionKey string    `json:"encryptedCollectionKey"`
	EncryptedName          string    `json:"encryptedName"`
	NameNonce              string    `json:"nameNonce"`
	CanUpload              bool      `json:"canUpload"`
	CanDelete              bool      `json:"canDelete"`
	UploadQuotaBytes       *int64    `json:"uploadQuotaBytes"`
	CreatedAt              time.Time `json:"createdAt"`
}

// POST /api/fed-proxy/incoming  — add incoming share by pasting invite URL
// @Summary      Accept a federated share invite
// @Tags         Federation Proxy
// @Accept       json
// @Produce      json
// @Security     BearerAuth
// @Param        body  body      AddIncomingShareRequest  true  "Invite URL"
// @Success      201   {object}  IncomingShare
// @Failure      400   {object}  ErrorResponse
// @Failure      401   {object}  ErrorResponse
// @Failure      502   {object}  ErrorResponse
// @Router       /fed-proxy/incoming [post]
func (h *FedProxyHandler) AddIncomingShare(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	var req struct {
		InviteURL string `json:"inviteUrl"` // https://server-b.com/invite/{token}
	}
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid request"})
	}
	if req.InviteURL == "" {
		return c.Status(400).JSON(fiber.Map{"error": "inviteUrl required"})
	}

	// Parse: https://server-b.com/invite/{token}
	// URL format: {scheme}://{host}/invite/{token}
	u := req.InviteURL
	idx := strings.Index(u, "/invite/")
	if idx < 0 {
		return c.Status(400).JSON(fiber.Map{"error": "invalid invite URL"})
	}
	remoteServer := u[:idx]
	token := u[idx+len("/invite/"):]
	if remoteServer == "" || token == "" {
		return c.Status(400).JSON(fiber.Map{"error": "invalid invite URL"})
	}

	// S1-4: Validate remote server URL to prevent SSRF
	allowHTTP := h.AppEnv != "production"
	if err := utils.ValidateFederationURL(remoteServer, allowHTTP); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid server URL: " + err.Error()})
	}

	// Fetch invite from remote server
	fetchURL := fmt.Sprintf("%s/api/fed/invites/%s", remoteServer, token)
	resp, err := fedHTTPClient.Get(fetchURL) //nolint:gosec — URL validated above
	if err != nil || resp.StatusCode != 200 {
		return c.Status(502).JSON(fiber.Map{"error": "failed to fetch invite from remote server"})
	}
	defer resp.Body.Close()

	var inviteData struct {
		WrappedKey       string `json:"wrappedKey"`
		EncryptedName    string `json:"encryptedName"`
		NameNonce        string `json:"nameNonce"`
		CanUpload        bool   `json:"canUpload"`
		CanDelete        bool   `json:"canDelete"`
		UploadQuotaBytes *int64 `json:"uploadQuotaBytes"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&inviteData); err != nil {
		return c.Status(502).JSON(fiber.Map{"error": "invalid invite data"})
	}

	// Store in federated_incoming_shares
	var shareID string
	err = h.DB.QueryRow(context.Background(), `
		INSERT INTO federated_incoming_shares (user_id, remote_server, remote_access_token,
		    encrypted_collection_key, encrypted_name, name_nonce,
		    can_upload, can_delete, upload_quota_bytes)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
		ON CONFLICT (user_id, remote_server, remote_access_token) DO UPDATE
		    SET encrypted_collection_key = EXCLUDED.encrypted_collection_key
		RETURNING id
	`, userID, remoteServer, token,
		inviteData.WrappedKey, inviteData.EncryptedName, inviteData.NameNonce,
		inviteData.CanUpload, inviteData.CanDelete, inviteData.UploadQuotaBytes,
	).Scan(&shareID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	return c.Status(201).JSON(fiber.Map{
		"id":                     shareID,
		"remoteServer":           remoteServer,
		"encryptedCollectionKey": inviteData.WrappedKey,
		"encryptedName":          inviteData.EncryptedName,
		"nameNonce":              inviteData.NameNonce,
		"canUpload":              inviteData.CanUpload,
		"canDelete":              inviteData.CanDelete,
		"uploadQuotaBytes":       inviteData.UploadQuotaBytes,
	})
}

// GET /api/fed-proxy/incoming  — list all incoming shares
// @Summary      List incoming federated shares
// @Tags         Federation Proxy
// @Produce      json
// @Security     BearerAuth
// @Success      200  {array}   IncomingShare
// @Failure      401  {object}  ErrorResponse
// @Router       /fed-proxy/incoming [get]
func (h *FedProxyHandler) ListIncomingShares(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	rows, err := h.DB.Query(context.Background(), `
		SELECT id, remote_server, encrypted_collection_key, encrypted_name, name_nonce,
		       can_upload, can_delete, upload_quota_bytes, created_at
		FROM federated_incoming_shares WHERE user_id = $1 ORDER BY created_at ASC
	`, userID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	defer rows.Close()

	var shares []IncomingShare
	for rows.Next() {
		var s IncomingShare
		if err := rows.Scan(&s.ID, &s.RemoteServer, &s.EncryptedCollectionKey,
			&s.EncryptedName, &s.NameNonce, &s.CanUpload, &s.CanDelete,
			&s.UploadQuotaBytes, &s.CreatedAt); err != nil {
			continue
		}
		shares = append(shares, s)
	}
	if shares == nil {
		shares = []IncomingShare{}
	}
	return c.JSON(shares)
}

// DELETE /api/fed-proxy/incoming/{shareId}  — remove incoming share
// @Summary      Remove an incoming federated share
// @Tags         Federation Proxy
// @Security     BearerAuth
// @Param        shareId  path  string  true  "Share UUID"
// @Success      204
// @Failure      401  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Router       /fed-proxy/incoming/{shareId} [delete]
func (h *FedProxyHandler) RemoveIncomingShare(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	shareID := c.Params("shareId")
	result, err := h.DB.Exec(context.Background(),
		`DELETE FROM federated_incoming_shares WHERE id = $1 AND user_id = $2`,
		shareID, userID)
	if err != nil || result.RowsAffected() == 0 {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}
	return c.SendStatus(204)
}

// GET /api/fed-proxy/{shareId}/files  — proxy list files
// @Summary      List files in a federated share (proxied)
// @Tags         Federation Proxy
// @Produce      json
// @Security     BearerAuth
// @Param        shareId  path      string  true  "Incoming share UUID"
// @Success      200      {array}   FileRow
// @Failure      401      {object}  ErrorResponse
// @Failure      404      {object}  ErrorResponse
// @Failure      502      {object}  ErrorResponse
// @Router       /fed-proxy/{shareId}/files [get]
func (h *FedProxyHandler) ProxyListFiles(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	shareID := c.Params("shareId")
	remoteServer, remoteToken, err := h.getShare(context.Background(), shareID, userID)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "share not found"})
	}
	url := fmt.Sprintf("%s/api/fed/shares/%s/files", remoteServer, remoteToken)
	resp, err := fedHTTPClient.Get(url) //nolint:gosec — remoteServer validated at insert time
	if err != nil {
		return c.Status(502).JSON(fiber.Map{"error": "remote error"})
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	c.Set("Content-Type", "application/json")
	return c.Status(resp.StatusCode).Send(body)
}

// GET /api/fed-proxy/{shareId}/files/{fileId}/download  — proxy download
// @Summary      Download a file from a federated share (proxied)
// @Tags         Federation Proxy
// @Produce      octet-stream
// @Security     BearerAuth
// @Param        shareId  path  string  true  "Incoming share UUID"
// @Param        fileId   path  string  true  "File UUID"
// @Success      200
// @Failure      401  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Failure      502  {object}  ErrorResponse
// @Router       /fed-proxy/{shareId}/files/{fileId}/download [get]
func (h *FedProxyHandler) ProxyDownload(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	shareID := c.Params("shareId")
	fileID := c.Params("fileId")
	remoteServer, remoteToken, err := h.getShare(context.Background(), shareID, userID)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "share not found"})
	}
	url := fmt.Sprintf("%s/api/fed/shares/%s/files/%s/download", remoteServer, remoteToken, fileID)
	resp, err := fedHTTPClient.Get(url) //nolint:gosec — remoteServer validated at insert time
	if err != nil || resp.StatusCode != 200 {
		return c.Status(502).JSON(fiber.Map{"error": "remote error"})
	}
	// Do NOT defer resp.Body.Close() — fasthttp reads stream lazily
	c.Set("Content-Type", "application/octet-stream")
	if cl := resp.Header.Get("Content-Length"); cl != "" {
		c.Set("Content-Length", cl)
	}
	return c.SendStream(resp.Body, -1)
}

// POST /api/fed-proxy/{shareId}/upload  — proxy upload
// @Summary      Upload a file to a federated share (proxied)
// @Tags         Federation Proxy
// @Accept       mpfd
// @Produce      json
// @Security     BearerAuth
// @Param        shareId            path      string  true  "Incoming share UUID"
// @Param        encryptedMetadata  formData  string  true  "Encrypted metadata (base64)"
// @Param        metadataNonce      formData  string  true  "Metadata nonce (base64)"
// @Param        encryptedFileKey   formData  string  true  "Encrypted file key (base64)"
// @Param        fileKeyNonce       formData  string  true  "File key nonce (base64)"
// @Param        file               formData  file    true  "Encrypted file content"
// @Success      201  {object}  MessageResponse
// @Failure      401  {object}  ErrorResponse
// @Failure      403  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Failure      502  {object}  ErrorResponse
// @Router       /fed-proxy/{shareId}/upload [post]
func (h *FedProxyHandler) ProxyUpload(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	shareID := c.Params("shareId")
	remoteServer, remoteToken, err := h.getShare(context.Background(), shareID, userID)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "share not found"})
	}
	url := fmt.Sprintf("%s/api/fed/shares/%s/files", remoteServer, remoteToken)

	// Forward the raw multipart body
	body := c.Body()
	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	req.Header.Set("Content-Type", c.Get("Content-Type"))
	resp, err := fedHTTPClient.Do(req)
	if err != nil {
		return c.Status(502).JSON(fiber.Map{"error": "remote error"})
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	c.Set("Content-Type", "application/json")
	return c.Status(resp.StatusCode).Send(respBody)
}

// DELETE /api/fed-proxy/{shareId}/files/{fileId}  — proxy delete
// @Summary      Delete a file in a federated share (proxied)
// @Tags         Federation Proxy
// @Security     BearerAuth
// @Param        shareId  path  string  true  "Incoming share UUID"
// @Param        fileId   path  string  true  "File UUID"
// @Success      204
// @Failure      401  {object}  ErrorResponse
// @Failure      404  {object}  ErrorResponse
// @Failure      502  {object}  ErrorResponse
// @Router       /fed-proxy/{shareId}/files/{fileId} [delete]
func (h *FedProxyHandler) ProxyDelete(c *fiber.Ctx) error {
	userID := middleware.UserID(c)
	shareID := c.Params("shareId")
	fileID := c.Params("fileId")
	remoteServer, remoteToken, err := h.getShare(context.Background(), shareID, userID)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "share not found"})
	}
	url := fmt.Sprintf("%s/api/fed/shares/%s/files/%s", remoteServer, remoteToken, fileID)
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	resp, err := fedHTTPClient.Do(req)
	if err != nil {
		return c.Status(502).JSON(fiber.Map{"error": "remote error"})
	}
	defer resp.Body.Close()
	return c.SendStatus(resp.StatusCode)
}

func (h *FedProxyHandler) getShare(ctx context.Context, shareID, userID string) (remoteServer, remoteToken string, err error) {
	err = h.DB.QueryRow(ctx,
		`SELECT remote_server, remote_access_token FROM federated_incoming_shares WHERE id = $1 AND user_id = $2`,
		shareID, userID,
	).Scan(&remoteServer, &remoteToken)
	return
}

File: crates/kutup-server/src/handlers/fedproxy.rs
//! Federation proxy endpoints — mirrors `backend/handlers/fedproxy.go`.
//!
//! These run on the *recipient's* server. The browser can't call a remote kutup server
//! directly (CORS, and it must not learn the remote access token), so it calls these
//! authenticated endpoints, which proxy to the remote `/api/fed/*` routes using the stored
//! per-share token. All outbound URLs are SSRF-validated (at accept time) and the shared
//! `FED_CLIENT` never follows redirects.

use axum::body::{Body, Bytes};
use axum::extract::{Path, State};
use axum::http::{header, HeaderValue, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use utoipa::ToSchema;
use uuid::Uuid;

use crate::error::{AppError, AppResult};
use crate::handlers::FED_CLIENT;
use crate::middleware::AuthUser;
use crate::{ssrf, AppState};

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct AddIncomingShareRequest {
    /// e.g. `https://server-b.com/invite/{token}`.
    invite_url: String,
}

/// What the remote `/api/fed/invites/{token}` returns.
#[derive(Debug, Deserialize)]
struct InviteData {
    #[serde(rename = "wrappedKey", default)]
    wrapped_key: String,
    #[serde(rename = "encryptedName", default)]
    encrypted_name: String,
    #[serde(rename = "nameNonce", default)]
    name_nonce: String,
    #[serde(rename = "canUpload", default)]
    can_upload: bool,
    #[serde(rename = "canDelete", default)]
    can_delete: bool,
    #[serde(rename = "uploadQuotaBytes", default)]
    upload_quota_bytes: Option<i64>,
}

/// The `AddIncomingShare` response (keys alphabetical, matching Go's marshalled `fiber.Map`).
#[derive(Debug, Serialize, ToSchema)]
struct AddIncomingShareResponse {
    #[serde(rename = "canDelete")]
    can_delete: bool,
    #[serde(rename = "canUpload")]
    can_upload: bool,
    #[serde(rename = "encryptedCollectionKey")]
    encrypted_collection_key: String,
    #[serde(rename = "encryptedName")]
    encrypted_name: String,
    id: Uuid,
    #[serde(rename = "nameNonce")]
    name_nonce: String,
    #[serde(rename = "remoteServer")]
    remote_server: String,
    #[serde(rename = "uploadQuotaBytes")]
    upload_quota_bytes: Option<i64>,
}

/// `POST /api/fed-proxy/incoming` — mirrors `AddIncomingShare`. Parses the invite URL,
/// SSRF-validates the remote host, fetches the invite, and stores it.
#[utoipa::path(
    post,
    path = "/api/fed-proxy/incoming",
    tag = "federation",
    security(("BearerAuth" = [])),
    request_body = crate::models::AddIncomingShareRequest,
    responses((status = 201, description = "Incoming share accepted", body = AddIncomingShareResponse))
)]
pub async fn add_incoming_share(
    State(state): State<AppState>,
    user: AuthUser,
    Json(req): Json<AddIncomingShareRequest>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    if req.invite_url.is_empty() {
        return Err(AppError::bad_request("inviteUrl required"));
    }

    // Split {scheme}://{host}/invite/{token}.
    let Some(idx) = req.invite_url.find("/invite/") else {
        return Err(AppError::bad_request("invalid invite URL"));
    };
    let remote_server = &req.invite_url[..idx];
    let token = &req.invite_url[idx + "/invite/".len()..];
    if remote_server.is_empty() || token.is_empty() {
        return Err(AppError::bad_request("invalid invite URL"));
    }

    let allow_http = state.config.app_env != "production";
    if let Err(e) = ssrf::validate_federation_url(remote_server, allow_http).await {
        return Err(AppError::bad_request(format!("invalid server URL: {e}")));
    }

    let fetch_url = format!("{remote_server}/api/fed/invites/{token}");
    let resp = FED_CLIENT.get(&fetch_url).send().await;
    let invite: InviteData = match resp {
        Ok(r) if r.status().as_u16() == 200 => match r.json().await {
            Ok(d) => d,
            Err(_) => {
                return Err(AppError::new(
                    StatusCode::BAD_GATEWAY,
                    "invalid invite data",
                ))
            }
        },
        _ => {
            return Err(AppError::new(
                StatusCode::BAD_GATEWAY,
                "failed to fetch invite from remote server",
            ))
        }
    };

    let share_id: Uuid = sqlx::query_scalar(
        r#"INSERT INTO federated_incoming_shares (user_id, remote_server, remote_access_token,
               encrypted_collection_key, encrypted_name, name_nonce,
               can_upload, can_delete, upload_quota_bytes)
           VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
           ON CONFLICT (user_id, remote_server, remote_access_token) DO UPDATE
               SET encrypted_collection_key = EXCLUDED.encrypted_collection_key
           RETURNING id"#,
    )
    .bind(user_id)
    .bind(remote_server)
    .bind(token)
    .bind(&invite.wrapped_key)
    .bind(&invite.encrypted_name)
    .bind(&invite.name_nonce)
    .bind(invite.can_upload)
    .bind(invite.can_delete)
    .bind(invite.upload_quota_bytes)
    .fetch_one(&state.pool)
    .await
    .map_err(|_| AppError::internal("internal error"))?;

    Ok((
        StatusCode::CREATED,
        Json(AddIncomingShareResponse {
            can_delete: invite.can_delete,
            can_upload: invite.can_upload,
            encrypted_collection_key: invite.wrapped_key,
            encrypted_name: invite.encrypted_name,
            id: share_id,
            name_nonce: invite.name_nonce,
            remote_server: remote_server.to_string(),
            upload_quota_bytes: invite.upload_quota_bytes,
        }),
    )
        .into_response())
}

/// An incoming federated share. Field order mirrors the Go struct.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
struct IncomingShare {
    id: Uuid,
    remote_server: String,
    encrypted_collection_key: String,
    encrypted_name: String,
    name_nonce: String,
    can_upload: bool,
    can_delete: bool,
    upload_quota_bytes: Option<i64>,
    #[serde(with = "time::serde::rfc3339")]
    created_at: OffsetDateTime,
}

/// `GET /api/fed-proxy/incoming` — mirrors `ListIncomingShares`.
#[utoipa::path(
    get,
    path = "/api/fed-proxy/incoming",
    tag = "federation",
    security(("BearerAuth" = [])),
    responses((status = 200, description = "The caller's incoming federated shares", body = Vec<IncomingShare>))
)]
pub async fn list_incoming_shares(
    State(state): State<AppState>,
    user: AuthUser,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    type IncomingTuple = (
        Uuid,
        String,
        String,
        String,
        String,
        bool,
        bool,
        Option<i64>,
        OffsetDateTime,
    );
    let rows: Vec<IncomingTuple> = sqlx::query_as(
        r#"SELECT id, remote_server, encrypted_collection_key, encrypted_name, name_nonce,
                  can_upload, can_delete, upload_quota_bytes, created_at
           FROM federated_incoming_shares WHERE user_id = $1 ORDER BY created_at ASC"#,
    )
    .bind(user_id)
    .fetch_all(&state.pool)
    .await
    .map_err(|_| AppError::internal("internal error"))?;

    let shares: Vec<IncomingShare> = rows
        .into_iter()
        .map(|(id, rs, eck, en, nn, cu, cd, uq, created)| IncomingShare {
            id,
            remote_server: rs,
            encrypted_collection_key: eck,
            encrypted_name: en,
            name_nonce: nn,
            can_upload: cu,
            can_delete: cd,
            upload_quota_bytes: uq,
            created_at: created,
        })
        .collect();
    Ok(Json(shares).into_response())
}

/// `DELETE /api/fed-proxy/incoming/{shareId}` — mirrors `RemoveIncomingShare`.
#[utoipa::path(
    delete,
    path = "/api/fed-proxy/incoming/{shareId}",
    tag = "federation",
    security(("BearerAuth" = [])),
    params(("shareId" = String, Path, description = "Incoming-share id")),
    responses((status = 204, description = "Incoming share removed"))
)]
pub async fn remove_incoming_share(
    State(state): State<AppState>,
    user: AuthUser,
    Path(share_id): Path<String>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    let sid = Uuid::parse_str(&share_id).map_err(|_| AppError::not_found("not found"))?;
    let res = sqlx::query("DELETE FROM federated_incoming_shares WHERE id = $1 AND user_id = $2")
        .bind(sid)
        .bind(user_id)
        .execute(&state.pool)
        .await;
    match res {
        Ok(r) if r.rows_affected() > 0 => Ok(StatusCode::NO_CONTENT.into_response()),
        _ => Err(AppError::not_found("not found")),
    }
}

/// Looks up the remote server + token for an incoming share — mirrors `getShare`.
async fn get_share(state: &AppState, share_id: &str, user_id: Uuid) -> Option<(String, String)> {
    let sid = Uuid::parse_str(share_id).ok()?;
    sqlx::query_as(
        "SELECT remote_server, remote_access_token FROM federated_incoming_shares WHERE id = $1 AND user_id = $2",
    )
    .bind(sid)
    .bind(user_id)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten()
}

/// `GET /api/fed-proxy/{shareId}/files` — mirrors `ProxyListFiles`.
#[utoipa::path(
    get,
    path = "/api/fed-proxy/{shareId}/files",
    tag = "federation",
    security(("BearerAuth" = [])),
    params(("shareId" = String, Path, description = "Incoming-share id")),
    responses((status = 200, description = "The remote share's file list, proxied verbatim (JSON)"))
)]
pub async fn proxy_list_files(
    State(state): State<AppState>,
    user: AuthUser,
    Path(share_id): Path<String>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    let Some((remote, token)) = get_share(&state, &share_id, user_id).await else {
        return Err(AppError::not_found("share not found"));
    };
    let url = format!("{remote}/api/fed/shares/{token}/files");
    let resp = FED_CLIENT
        .get(&url)
        .send()
        .await
        .map_err(|_| AppError::new(StatusCode::BAD_GATEWAY, "remote error"))?;
    let status = StatusCode::from_u16(resp.status().as_u16()).unwrap_or(StatusCode::BAD_GATEWAY);
    let body = resp.bytes().await.unwrap_or_default();
    Ok((status, [(header::CONTENT_TYPE, "application/json")], body).into_response())
}

/// `GET /api/fed-proxy/{shareId}/files/{fileId}/download` — mirrors `ProxyDownload`. Streams
/// the remote response body straight back to the caller.
#[utoipa::path(
    get,
    path = "/api/fed-proxy/{shareId}/files/{fileId}/download",
    tag = "federation",
    security(("BearerAuth" = [])),
    params(
        ("shareId" = String, Path, description = "Incoming-share id"),
        ("fileId" = String, Path, description = "Remote file id")
    ),
    responses((status = 200, description = "The encrypted blob streamed from the remote server (application/octet-stream)"))
)]
pub async fn proxy_download(
    State(state): State<AppState>,
    user: AuthUser,
    Path((share_id, file_id)): Path<(String, String)>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    let Some((remote, token)) = get_share(&state, &share_id, user_id).await else {
        return Err(AppError::not_found("share not found"));
    };
    let url = format!("{remote}/api/fed/shares/{token}/files/{file_id}/download");
    let resp = match FED_CLIENT.get(&url).send().await {
        Ok(r) if r.status().as_u16() == 200 => r,
        _ => return Err(AppError::new(StatusCode::BAD_GATEWAY, "remote error")),
    };
    let content_length = resp
        .headers()
        .get(header::CONTENT_LENGTH)
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string());
    let mut builder = Response::builder().header(header::CONTENT_TYPE, "application/octet-stream");
    if let Some(cl) = content_length {
        if let Ok(v) = HeaderValue::from_str(&cl) {
            builder = builder.header(header::CONTENT_LENGTH, v);
        }
    }
    let stream = resp.bytes_stream();
    Ok(builder
        .body(Body::from_stream(stream))
        .expect("valid proxy-download response"))
}

/// `POST /api/fed-proxy/{shareId}/upload` — mirrors `ProxyUpload`. Forwards the raw multipart
/// body (with its Content-Type) to the remote share.
#[utoipa::path(
    post,
    path = "/api/fed-proxy/{shareId}/upload",
    tag = "federation",
    security(("BearerAuth" = [])),
    params(("shareId" = String, Path, description = "Incoming-share id")),
    request_body(
        content = Vec<u8>,
        content_type = "multipart/form-data",
        description = "Raw multipart body forwarded verbatim to the remote share"
    ),
    responses((status = 201, description = "The remote upload result, proxied verbatim (JSON)"))
)]
pub async fn proxy_upload(
    State(state): State<AppState>,
    user: AuthUser,
    Path(share_id): Path<String>,
    headers: axum::http::HeaderMap,
    body: Bytes,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    let Some((remote, token)) = get_share(&state, &share_id, user_id).await else {
        return Err(AppError::not_found("share not found"));
    };
    let url = format!("{remote}/api/fed/shares/{token}/files");
    let content_type = headers
        .get(header::CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("application/octet-stream")
        .to_string();
    let resp = FED_CLIENT
        .post(&url)
        .header(header::CONTENT_TYPE, content_type)
        .body(body.to_vec())
        .send()
        .await
        .map_err(|_| AppError::new(StatusCode::BAD_GATEWAY, "remote error"))?;
    let status = StatusCode::from_u16(resp.status().as_u16()).unwrap_or(StatusCode::BAD_GATEWAY);
    let rbody = resp.bytes().await.unwrap_or_default();
    Ok((status, [(header::CONTENT_TYPE, "application/json")], rbody).into_response())
}

/// `DELETE /api/fed-proxy/{shareId}/files/{fileId}` — mirrors `ProxyDelete`.
#[utoipa::path(
    delete,
    path = "/api/fed-proxy/{shareId}/files/{fileId}",
    tag = "federation",
    security(("BearerAuth" = [])),
    params(
        ("shareId" = String, Path, description = "Incoming-share id"),
        ("fileId" = String, Path, description = "Remote file id")
    ),
    responses((status = 204, description = "Deleted on the remote server (status proxied)"))
)]
pub async fn proxy_delete(
    State(state): State<AppState>,
    user: AuthUser,
    Path((share_id, file_id)): Path<(String, String)>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;
    let Some((remote, token)) = get_share(&state, &share_id, user_id).await else {
        return Err(AppError::not_found("share not found"));
    };
    let url = format!("{remote}/api/fed/shares/{token}/files/{file_id}");
    let resp = FED_CLIENT
        .delete(&url)
        .send()
        .await
        .map_err(|_| AppError::new(StatusCode::BAD_GATEWAY, "remote error"))?;
    let status = StatusCode::from_u16(resp.status().as_u16()).unwrap_or(StatusCode::BAD_GATEWAY);
    Ok(status.into_response())
}
